File: nginx/src/script/object/ball.js
import * as THREE from 'three';

export class Ball {
	constructor() {
		const geometry = new THREE.SphereGeometry(0.1, 32, 16);
		const material = new THREE.MeshBasicMaterial({ color: 0xFFE400 });
		this.mesh = new THREE.Mesh(geometry, material);
		this.velocity = new THREE.Vector3(0, 0, 0);
		this.isMoving = false;

		const random = Math.random() * 2 - 1;
		const direction = random > 0 ? 1 : -1;
		this.reset(direction);
	}
	reset(direction) {
		const randomY = Math.random() * 6.6 - 3.3;
		this.mesh.position.set(0, randomY, 0);

		this.velocity.set(0, 0, 0);
		this.isMoving = false;
		setTimeout(() => this.shoot(direction), 1000);
	}
	shoot(direction) {
		const angles = [45, 315, 135, 225];
		let index = Math.floor(Math.random() * 2);
		if (direction > 0) {
			index += 2;
		}
		const radian = angles[index] * Math.PI / 180;
		const speed = 0.3;
		this.velocity.set(
			Math.cos(radian) * speed,
			Math.sin(radian) * speed,
			0
		);
		this.isMoving = true;
	}
	move(player1, player2, deltatime) {
		if (!this.isMoving) {
			return false;
		}
		const limit = 3.3;
		const speed = 30;
		this.mesh.position.add(this.velocity.clone().multiplyScalar(speed * deltatime));
		if (Math.abs(this.mesh.position.x) > 7) {
			let direction = 1;
			if (this.mesh.position.x < 0) {
				direction = -1;
				player2.score++;
			} else {
				player1.score++;
			}
			this.reset(direction);
			return true;
		}
		if (Math.abs(this.mesh.position.y) > limit) {
			this.velocity.y *= -1;
			this.mesh.position.y = Math.sign(this.mesh.position.y) * limit;
		}
		this.checkCollision(player1);
		this.checkCollision(player2);
	}
	checkCollision(player) {
		const radius = 0.1;
		const box = player.getBound();
		if (this.mesh.position.x - radius < box.maxX && 
			this.mesh.position.x + radius > box.minX && 
			this.mesh.position.y + radius > box.minY && 
			this.mesh.position.y - radius < box.maxY) {
			this.velocity.x *= -1;
			if (this.velocity.x > 0) {
				this.mesh.position.x = box.maxX + radius;
			} else {
				this.mesh.position.x = box.minX - radius;
			}
			const hitPoint = (this.mesh.position.y - player.mesh.position.y) / (box.maxY - box.minY);
			this.velocity.y = hitPoint * 1;
			const currentSpeed = this.velocity.length();
			this.velocity.normalize().multiplyScalar(currentSpeed);
		}
	}
}

File: django/src/game/ball.py
import random
import math
from django.conf import settings
from .player import Player

class Ball:
    def __init__(self):
        self._radius = settings.BALL_RADIUS
        self._position = {"x": 0, "y": 0}
        self._velocity = {"x": 0, "y": 0}
        self._isMoving = False
        self._direction = random.random() * 2 - 1 > 0
        self.shoot()
        
    def reset(self):
        randomY = random.random() * 6.6 - 3.3
        self._position = {"x": 0, "y": randomY}
        self._velocity = {"x": 0, "y": 0}
        self._isMoving = False
        self.shoot()
    
    def shoot(self):
        angles = [45, 315, 135, 225]
        index = math.floor(random.random() * 2)
        if self._direction:
            index += 2
                     
        radian = angles[index] * math.pi / 180
        speed = settings.SPEED_START
        self._velocity["x"] = math.cos(radian) * speed
        self._velocity["y"] = math.sin(radian) * speed
        self._isMoving = True
        
    def move(self, player1: Player, player2: Player, deltatime):
        if not self._isMoving:
            return False
        
        limit = settings.HEIGHT_LIMIT
        speed = 30
        self._position["x"] += self._velocity["x"] * (speed * deltatime)
        self._position["y"] += self._velocity["y"] * (speed * deltatime)
        
        if abs(self._position["x"]) > 7:
            self._direction = True
            if self._position["x"] < 0:
                self._direction = False
                player2.score += 1
            else:
                player1.score += 1
            self.reset()
            return True
        
        if abs(self._position["y"]) > limit:
            self._velocity["y"] *= -1
            self._position["y"] = math.copysign(1, self._position["y"]) * limit
            
        self.checkCollision(player1)
        self.checkCollision(player2)
        return False

    def checkCollision(self, player: Player):
        box = player.getBound()
        if self._position["x"] - self._radius < box["maxX"] and \
            self._position["x"] + self._radius > box["minX"] and \
            self._position["y"] + self._radius > box["minY"] and \
            self._position["y"] - self._radius < box["maxY"]:
            
            self._velocity["x"] *= -1
            if self._velocity["x"] > 0:
                self._position["x"] = box["maxX"] + self._radius
            else:
                self._position["x"] = box["minX"] - self._radius
            
            hitPoint = (self._position["y"] - player.position["y"]) / (box["maxY"] - box["minY"])
            self._velocity["y"] = hitPoint * 1
            currentSpeed = math.dist((0, 0, 0), (self._velocity["x"], self._velocity["y"], 0))
            length = math.sqrt(math.pow(self._velocity["x"], 2) + math.pow(self._velocity["y"], 2))
            self._velocity["x"] = self._velocity["x"] / length * currentSpeed
            self._velocity["y"] = self._velocity["y"] / length * currentSpeed

    def getJson(self):
        return {
            "position": self._position
        }